tests/conftest.py

```diff
@@ -1,80 +1,126 @@
-    # types.h should be in src/types.h
-    types_h_path = conftest_folder_path.parent / "src" / "transaction" / "types.h"
+@pytest.fixture(scope="session")
+def backend_name(pytestconfig):
+    return pytestconfig.getoption("backend")
 
-    if not types_h_path.is_file():
-        raise FileNotFoundError(f"Can't find types.h: '{types_h_path}'")
 
-    return types_h_path
+@pytest.fixture(scope="session")
+def display(pytestconfig):
+    return pytestconfig.getoption("display")
 
 
 @pytest.fixture(scope="session")
-def hid(pytestconfig):
-    return pytestconfig.getoption("hid")
+def golden_run(pytestconfig):
+    return pytestconfig.getoption("golden_run")
+
+
+@pytest.fixture
+def test_name(request):
+    # Get the name of current pytest test
+    test_name = request.node.name
+
+    # Remove firmware suffix:
+    # -  test_xxx_transaction_ok[nanox 2.0.2]
+    # => test_xxx_transaction_ok
+    return test_name.split("[")[0]
+
+
+# Glue to call every test that depends on the firmware once for each required firmware
+def pytest_generate_tests(metafunc):
+    if "firmware" in metafunc.fixturenames:
+        fw_list = []
+        ids = []
+        # First pass: enable only demanded firmwares
+        for fw in FIRMWARES:
+            if metafunc.config.getoption(fw.device):
+                fw_list.append(fw)
+                ids.append(fw.device + " " + fw.version)
+        # Second pass if no specific firmware demanded: add them all
+        if not fw_list:
+            for fw in FIRMWARES:
+                fw_list.append(fw)
+                ids.append(fw.device + " " + fw.version)
+        metafunc.parametrize("firmware", fw_list, ids=ids)
+
+
+def prepare_speculos_args(firmware: Firmware, display: bool):
+    speculos_args = []
+
+    if display:
+        speculos_args += ["--display", "qt"]
+
+    app_path = app_path_from_app_name(APPS_DIRECTORY, APP_NAME, firmware.device)
+
+    return ([app_path], {"args": speculos_args})
+
+
+# Depending on the "--backend" option value, a different backend is
+# instantiated, and the tests will either run on Speculos or on a physical
+# device depending on the backend
+def create_backend(backend_name: str, firmware: Firmware, display: bool):
+    if backend_name.lower() == "ledgercomm":
+        return LedgerCommBackend(firmware, interface="hid")
+    elif backend_name.lower() == "ledgerwallet":
+        return LedgerWalletBackend(firmware)
+    elif backend_name.lower() == "speculos":
+        args, kwargs = prepare_speculos_args(firmware, display)
+        return SpeculosBackend(*args, firmware, **kwargs)
+    else:
+        raise ValueError(f"Backend '{backend_name}' is unknown. Valid backends are: {BACKENDS}")
 
 
-@pytest.fixture(scope="session")
-def headless(pytestconfig):
-    return pytestconfig.getoption("headless")
+# This final fixture will return the properly configured backend, to be used in tests
+@pytest.fixture
+def backend(backend_name, firmware, display):
```

Comment:
Should we rename it backend also on the ragger template?
I think we should :)

Reply:
will do next